[AMDGPU][MachineVerifier] test failures in SIFoldOperands

AbhayKanhere opened the pull request:
After PR:https://github.com/llvm/llvm-project/pull/151421 merged following fails in SIFoldOperands showed up.

LLVM :: CodeGen/AMDGPU/GlobalISel/llvm.amdgcn.mfma.gfx90a.ll
LLVM :: CodeGen/AMDGPU/llvm.amdgcn.mfma.gfx90a.ll
LLVM :: CodeGen/AMDGPU/llvm.amdgcn.mfma.ll
LLVM :: CodeGen/AMDGPU/mfma-loop.ll
LLVM :: CodeGen/AMDGPU/rewrite-vgpr-mfma-to-agpr.ll

In Folding code, if folded operand is register ensure earlyClobber is set.

AbhayKanhere committed:
[AMDGPU][MachineVerifier] test failures in SIFoldOperands

After PR:https://github.com/llvm/llvm-project/pull/151421 merged
following fails in SIFoldOperands showed up.

LLVM :: CodeGen/AMDGPU/GlobalISel/llvm.amdgcn.mfma.gfx90a.ll
LLVM :: CodeGen/AMDGPU/llvm.amdgcn.mfma.gfx90a.ll
LLVM :: CodeGen/AMDGPU/llvm.amdgcn.mfma.ll
LLVM :: CodeGen/AMDGPU/mfma-loop.ll
LLVM :: CodeGen/AMDGPU/rewrite-vgpr-mfma-to-agpr.ll

In Folding code, if folded operand is register ensure earlyClobber
is set.

AbhayKanhere committed:
Merge branch 'main' into eng/abhay/machine-verify-early-clobber-fix-tests

shiltian commented on llvm/lib/Target/AMDGPU/SIFoldOperands.cpp:
```diff
@@ -689,6 +689,10 @@ bool SIFoldOperandsImpl::updateOperand(FoldCandidate &Fold) const {
+    const MCInstrDesc &MCID = MI->getDesc();
+    if (MCID.getOperandConstraint(OpNo, MCOI::EARLY_CLOBBER) != -1) {
+      MI->getOperand(OpNo).setIsEarlyClobber(true);
+    }
```
```suggestion
    if (MCID.getOperandConstraint(OpNo, MCOI::EARLY_CLOBBER) != -1)
      MI->getOperand(OpNo).setIsEarlyClobber(true);
```

shiltian reviewed (commented):
The fix looks reasonable to me, but I'm not sure if we want to add a test case w/o need of expensive checks.

arsenm committed:
Apply suggestion from @shiltian

Co-authored-by: Shilei Tian <i@tianshilei.me>

AbhayKanhere commented on llvm/lib/Target/AMDGPU/SIFoldOperands.cpp:
```diff
@@ -689,6 +689,9 @@ bool SIFoldOperandsImpl::updateOperand(FoldCandidate &Fold) const {
       return false;
 
+    const MCInstrDesc &MCID = MI->getDesc();
+    if (MCID.getOperandConstraint(OpNo, MCOI::EARLY_CLOBBER) != -1)
```
@shiltian @arsenm we can modify test to always check  -verify-machineinstrs - and looks like this fix is not sufficient. let me fix these tests by adding -verify-machineinstrs and then look for real fix. thanks!

AbhayKanhere replied:
SIFoldOperands appears to be the first place after which machine verifier runs ? ensuring the check for all definitions there appears to fix these tests. I had removed this code earlier during review. Modified it to iterate on all defs.
Also modified machine verifier Check to accurately inform which MachineOperand does not have earlyClobber. 
thanks!

AbhayKanhere committed:
Merge branch 'llvm:main' into eng/abhay/machine-verify-early-clobber-fix-tests

AbhayKanhere committed:
Set early clobber checking for all defs in SIFoldOperands.
Fixed tests - added -verify-machineinstrs so fails are visible sooner.
This resolves all fails so far.

AbhayKanhere committed:
Merge branch 'main' into eng/abhay/machine-verify-early-clobber-fix-tests

AbhayKanhere commented on llvm/lib/CodeGen/MachineVerifier.cpp:
```diff
@@ -2587,9 +2587,9 @@ MachineVerifier::visitMachineOperand(const MachineOperand *MO, unsigned MONum) {
   if (MCID.getOperandConstraint(MONum, MCOI::EARLY_CLOBBER) != -1) {
     if (!MO->isReg())
-      report("Early clobber must be a register", MI);
+      report("Early clobber must be a register", MO, MONum);
```
ok

AbhayKanhere committed:
review comment resolved
removed change in MachineVerifier

AbhayKanhere commented:
I would like to merge this once approved to unblock the fails in nightly runs. thanks!

AbhayKanhere committed:
undid one more instance of changes
